pkg/kafka/policy.go

```diff
@@ -281,9 +222,42 @@ func (req *RequestMessage) ruleMatches(rule api.PortRuleKafka) bool {
 // rules. The function will return true if the policy allows the message,
 // otherwise false is returned.
 func (req *RequestMessage) MatchesRule(rules []api.PortRuleKafka) bool {
+	reqTopicsMap := map[string]bool{}
+
+	// Maintain a map of all topics in the request
+	// Only if all the topics in a request are
+	// allowed by the rules, we should allow a request.
+	topics := req.GetTopics()
+	for _, topic := range topics {
+		reqTopicsMap[topic] = true
+	}
+
 	for _, rule := range rules {
-		if req.ruleMatches(rule) {
-			return true
+		switch len(topics) {
+		case 0:
+			if req.ruleMatches(rule, "") {
+				return true
+			}
+		default:
+			if rule.Topic != "" {
+				_, ok := reqTopicsMap[rule.Topic]
+				if ok && req.ruleMatches(rule, rule.Topic) {
```

Comment:
This doesn't make sense to me.
You set `matchTopic` to `rule.Topic`. So in every `match*Req` method, the test:
```
if rule.Topic != "" && rule.Topic != matchTopic {
```
is effectively:
```
if rule.Topic != "" && rule.Topic != rule.Topic {
```
which is never true.

In this PR, `matchTopic` is always set to either `rule.Topic` or `""`.
When it is set to `rule.Topic`, the test in those methods is a no-op, it's never true as I shown above.
When it is set to `""`, this is always when `rule.Topic == ""`, so this test is never true in this case too.

The `if rule.Topic != "" && rule.Topic != matchTopic {` tests in every method are therefore dead code now. Remove all of them.

And since you don't use the `matchTopic` parameters for anything else, just remove that parameter everywhere too.

Reply:
I was trying to keep the same semantics as before, but I agree, now with a hash map, and no topic iteration, this code is redundant. Let me fix that. thanks for the comment.

Reply:
done